pkg/uprobetracer/tracer.go

```diff
@@ -0,0 +1,326 @@
+	inodeRefCount map[inodeUUID]*inodeKeeper
+	// used as a set, keeps PIDs of the pending containers
+	pendingContainerPids map[uint32]bool
+
+	logger logger.Logger
+
+	closed bool
+	mu     sync.Mutex
+}
+
+func NewTracer[Event any](logger logger.Logger) (*Tracer[Event], error) {
+	t := &Tracer[Event]{
+		containerPid2Inodes:  make(map[uint32][]inodeUUID),
+		inodeRefCount:        make(map[inodeUUID]*inodeKeeper),
+		pendingContainerPids: make(map[uint32]bool),
+		logger:               logger,
+		closed:               false,
+	}
+	return t, nil
+}
+
+// AttachProg loads the ebpf program, and try attaching if there are pending containers
+func (t *Tracer[Event]) AttachProg(progName string, progType ProgType, attachTo string, prog *ebpf.Program) error {
+	if progType != ProgUprobe && progType != ProgUretprobe {
+		return fmt.Errorf("unsupported uprobe prog type: %q", progType)
+	}
+
+	if prog == nil {
+		return errors.New("prog does not exist")
+	}
+	if t.prog != nil {
+		return errors.New("loading uprobe program twice")
+	}
+
+	parts := strings.Split(attachTo, ":")
+	if len(parts) < 2 {
+		return fmt.Errorf("invalid section name %q", attachTo)
+	}
+	if !filepath.IsAbs(parts[0]) && strings.Contains(parts[0], "/") {
+		return fmt.Errorf("section name must be either an absolute path or a library name: %q", parts[0])
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if t.closed {
+		return errors.New("uprobetracer has been closed")
+	}
+
+	t.progName = progName
+	t.progType = progType
+	t.attachFilePath = parts[0]
+	t.attachSymbol = parts[1]
+	t.prog = prog
+
+	// attach to pending containers, then release the pending list
+	for pid := range t.pendingContainerPids {
+		t.attach(pid)
+	}
+	t.pendingContainerPids = nil
+
+	return nil
+}
+
+func searchForLibrary(containerPid uint32, filePath string) ([]string, error) {
+	var libraryPaths []string
+	var securedLibraryPaths []string
+
+	if !filepath.IsAbs(filePath) {
+		containerLdCachePath, err := securejoin.SecureJoin(filepath.Join(host.HostProcFs, fmt.Sprint(containerPid), "root"), "etc/ld.so.cache")
+		if err == nil {
+			ldCachePaths, err := parseLdCache(containerLdCachePath, filePath)
+			if err == nil {
+				libraryPaths = ldCachePaths
+			} else {
+				return nil, fmt.Errorf("parsing ld cache: %q", err.Error())
+			}
+		} else {
+			return nil, fmt.Errorf("path is not secure: %q", err.Error())
+		}
+	} else {
+		libraryPaths = append(libraryPaths, filePath)
+	}
+	for _, libraryPath := range libraryPaths {
+		securedLibraryPath, err := securejoin.SecureJoin(filepath.Join(host.HostProcFs, fmt.Sprint(containerPid), "root"), libraryPath)
+		if err == nil {
+			securedLibraryPaths = append(securedLibraryPaths, securedLibraryPath)
+		} else {
+			return nil, fmt.Errorf("path is not secure: %q", err.Error())
+		}
+	}
+	return securedLibraryPaths, nil
+}
+
+// attach ebpf program to a self-hosted inode
+func (t *Tracer[Event]) attachEbpf(file *os.File) (link.Link, error) {
```

Comment:
```suggestion
// attach uprobe program to the inode of the file passed in parameter
func (t *Tracer[Event]) attachUprobe(file *os.File) (link.Link, error) {
```

I suggest the name `attachUprobe`, I think it brings less confusion if it is a ebpf module or a ebpf program or a ebpf map.